File: solutions/cpp/0290.cpp
class Solution {
 public:
  bool wordPattern(string pattern, string str) {
    const int n = pattern.length();

    vector<int> charToIndex(26);
    unordered_map<string, int> stringToIndex;
    istringstream iss(str);

    int i = 0;
    for (string word; iss >> word; ++i) {
      if (i == n || charToIndex[pattern[i] - 'a'] != stringToIndex[word])
        return false;
      charToIndex[pattern[i] - 'a'] = i + 1;
      stringToIndex[word] = i + 1;
    }

    return i == n;
  }
};

File: solutions/java/0290.java
class Solution {
  public boolean wordPattern(String pattern, String str) {
    String[] words = str.split(" ");
    if (words.length != pattern.length())
      return false;

    Map<Character, Integer> charMap = new HashMap<>();
    Map<String, Integer> strMap = new HashMap<>();

    for (Integer i = 0; i < pattern.length(); ++i)
      if (charMap.put(pattern.charAt(i), i) != strMap.put(words[i], i))
        return false;

    return true;
  }
}